common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -146,138 +142,127 @@ trait MultipleReadersSingleWriterCache[W, Winfo] {
+                    // a pre-existing owner will take care of the invalidation
+                    invalidator
             }
-        } else invalidator // not caching
+        }
     }
 
-    /**
-     * This method may initiate a read from the backing store, and potentially stores the result in the cache.
-     */
-    protected def cacheLookup[Wsuper >: W](key: Any, generator: => Future[W], fromCache: Boolean = cacheEnabled)(
-        implicit ec: ExecutionContext, transid: TransactionId, logger: Logging): Future[W] = {
-        if (fromCache) {
-            val promise = Promise[W] // this promise completes with the generator value
-
-            // try inserting our desired entry...
-            val desiredEntry = Entry(transid, ReadInProgress, Some(promise.future))
-            cache(key)(desiredEntry) flatMap { actualEntry =>
-                // ... and see what we get back
-
-                actualEntry.state.get match {
-                    case Cached =>
-                        logger.debug(this, "cached read")
-                        makeNoteOfCacheHit(key)
-                        actualEntry.unpack
-
-                    case ReadInProgress =>
-                        if (actualEntry == desiredEntry) {
-                            logger.debug(this, "read initiated");
-                            makeNoteOfCacheMiss(key)
-                            // updating the cache with the new value is done in the listener
-                            // and will complete unless an invalidation request or an intervening
-                            // write occur in the meantime
-                            listenForReadDone(key, actualEntry, generator, promise)
-                            actualEntry.unpack
-                        } else {
-                            logger.debug(this, "coalesced read")
-                            makeNoteOfCacheHit(key)
-                            actualEntry.unpack
+    def cacheLookup[Wsub <: W](key: Any, generator: => Future[Wsub])(
+        implicit ec: ExecutionContext, transid: TransactionId, logger: Logging, mw: Manifest[Wsub]): Future[Wsub] = {
+        val promise = Promise[W] // this promise completes with the generator value
+
+        // try inserting our desired entry...
+        val desiredEntry = Entry(transid, ReadInProgress, Some(promise.future))
+        cache(key)(desiredEntry) flatMap { actualEntry =>
+            // ... and see what we get back
+
+            actualEntry.state.get match {
+                case Cached =>
+                    logger.debug(this, "cached read")
+                    makeNoteOfCacheHit(key)
+                    actualEntry.unpack.map { entry =>
+                        if (entry.getClass != mw.runtimeClass) {
+                            throw DocumentTypeMismatchException(s"document type ${entry.getClass} did not match expected type ${mw.runtimeClass}.")
                         }
+                        entry.asInstanceOf[Wsub]
```

Comment:
It's the same like getting the WhiskEntity out of the Datastore:
https://github.com/apache/incubator-openwhisk/blob/master/common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala#L148